miio/curtain_babai.py

```diff
@@ -0,0 +1,201 @@
+from enum import Enum
+from time import sleep
+from typing import Any, Dict, List
+
+import click
+
+from .click_common import EnumType, command, format_output
+from .exceptions import DeviceError
+from .miot_device import DeviceStatus, MiotDevice
+
+_MAPPING = {
+    # Source
+    # http://miot-spec.org/miot-spec-v2/instances?status=all
+    # https://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:curtain:0000A00C:babai-190812:1:0000C805
+    # Curtain
+    "motor_control": {"siid": 2, "piid": 1},  # 0 - Pause, 1 - Open, 2 - Close
+    "current_position": {"siid": 2, "piid": 2},  # Range: [0, 100, 1]
+    "target_position": {"siid": 2, "piid": 3},  # Range: [0, 100, 1]
+    "mode": {"siid": 2, "piid": 4},  # 0 - Normal,  1 - Reversal, 2 - Calibrate
+}
+
+# Model: "OnViz Curtain Controller (Wi-Fi)"
+MODEL_CURTAIN_BABAI = "babai.curtain.190812"
+
+
+class MotorControl(Enum):
+    Pause = 0
+    Open = 1
+    Close = 2
+
+
+class Mode(Enum):
+    Normal = 0
+    Reverse = 1
+    Calibrate = 2
+
+
+class CurtainStatus(DeviceStatus):
+    def __init__(self, data: Dict[str, Any]) -> None:
+        """Response from device.
+
+        {'motor_control': 0}
+        """
+        self.data = data
+
+    @property
+    def status(self) -> MotorControl:
+        """Device status."""
+        return MotorControl(self.data["motor_control"])
+
+    @property
+    def motor_control(self) -> MotorControl:
+        """Device status."""
+        return MotorControl(self.data["motor_control"])
+
+    @property
+    def is_open(self) -> bool:
+        """Device status."""
+        return self.motor_control == MotorControl.Open
+
+    @property
+    def is_close(self) -> bool:
+        """Device status."""
+        return self.motor_control == MotorControl.Close
+
+    @property
+    def mode(self) -> Mode:
+        """Motor rotation polarity."""
+        return Mode(self.data["mode"])
+
+    @property
+    def current_position(self) -> int:
+        """Run time of the motor."""
+        return self.data["current_position"]
+
+    @property
+    def target_position(self) -> int:
+        """Target curtain position."""
+        return self.data["target_position"]
+
+
+class CurtainBabai(MiotDevice):
+    """Main class representing the babai.curtain.190812 curtain."""
+
+    mapping = _MAPPING
+
+    @command(
+        default_output=format_output(
+            "",
+            "Motor control: {result.motor_control}\n"
+            "Motor mode: {result.mode}\n"
+            "Current position: {result.current_position}\n"
+            "Target position: {result.target_position}\n",
+        )
+    )
+    def status(self) -> CurtainStatus:
+        """Retrieve properties."""
+        # if use get_properties_for_mapping and max_len>1 then code -4004.
```

Comment:
So the device allows only requesting one property per request? If you define `max_properties=1`, is that not working in this case?

Reply:
{'code': -9999, 'message': 'user ack timeout'}